### ISSUE TITLE
Optimize adiabatically-contracted halo potential functions

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#109]
Commit 8784705a9b4084a1d6ff8ff9478185983662190e refactored the adiabatic contraction calculations. One improvement resulting from this is that the gravitational potential of adiabatically-contracted halos is now utilized. A downside of this is that the calculation is slow. Comparing commit 0028f0e2e7ec978526cb5f70f314d6c5273ef754 to commit 98c83da50a1a73d3d120294c63f0aa39599aee20 running under `callgrind` on the attached benchmark parameter file shows the following costs for the "`dark_matter_profile_potential_task()`" function:
```
0028f0e2e7ec978526cb5f70f314d6c5273ef754 : 1,264,325,300

98c83da50a1a73d3d120294c63f0aa39599aee20 : 18,978,484,446
```
Almost all of this extra time is used in the adiabatic contraction calculation, specifically in finding the initial radius using a root-finder.

**Comment by abensonca** [comment]
<?xml version="1.0" encoding="UTF-8"?>


<parameters>
  <formatVersion>2</formatVersion>
  <version>0.9.4</version>

  
  <treeNodeMethodBasic value="standard"/>
  <treeNodeMethodBlackHole value="standard"/>
  <treeNodeMethodDarkMatterProfile value="scale"/>
  <treeNodeMethodDisk value="standard"/>
  <treeNodeMethodHotHalo value="standard"/>
  <treeNodeMethodSatellite value="standard"/>
  <treeNodeMethodSpheroid value="standard"/>
  <diskMassDistribution value="exponentialDisk">
    <dimensionless value="true"/>
  </diskMassDistribution>
  <spheroidMassDistribution value="hernquist">
    <dimensionless value="true"/>
  </spheroidMassDistribution>
  <treeNodeMethodSpin value="random"/>

  
  <cosmologyFunctionsMethod value="matterLambda"/>
  <cosmologyParametersMethod value="simple">
    <HubbleConstant value="70.2"/>
    <OmegaMatter value="0.2725"/>
    <OmegaDarkEnergy value="0.7275"/>
    <OmegaBaryon value="0.0455"/>
    <temperatureCMB value="2.72548"/>
  </cosmologyParametersMethod>

  
  <transferFunctionMethod value="eisensteinHu1999">
    <neutrinoNumberEffective value="3.046"/>
    <neutrinoMassSummed value="0.000"/>
  </transferFunctionMethod>
  <powerSpectrumPrimordialMethod value="powerLaw">
    <index value="0.961"/>
    <wavenumberReference value="1.000"/>
    <running value="0.000"/>
  </powerSpectrumPrimordialMethod>
  <powerSpectrumPrimordialTransferredMethod value="simple"/>
  <cosmologicalMassVarianceMethod value="filteredPower">
      <sigma_8 value="0.807"/>
  </cosmologicalMassVarianceMethod>
  
  
  <linearGrowthMethod value="collisionlessMatter"/>
  <haloMassFunctionMethod value="tinker2008"/>
  <criticalOverdensityMethod value="sphericalCollapseClsnlssMttrCsmlgclCnstnt"/>
  <virialDensityContrastMethod value="sphericalCollapseClsnlssMttrCsmlgclCnstnt"/>

  
  <mergerTreeConstructorMethod value="build"/>
  <mergerTreeBuilderMethod value="cole2000">
    <accretionLimit value="0.1"/>
    <mergeProbability value="0.1"/>
  </mergerTreeBuilderMethod>
  <mergerTreeBranchingProbabilityMethod value="parkinsonColeHelly">
    <G0 value="+0.57"/>
    <gamma1 value="+0.38"/>
    <gamma2 value="-0.01"/>
    <accuracyFirstOrder value="+0.10"/>
  </mergerTreeBranchingProbabilityMethod>
  <mergerTreeBuildMassesMethod value="sampledDistributionUniform">
    <massTreeMinimum value="1.0e10"/>
    <massTreeMaximum value="1.0e13"/>
    <treesPerDecade value="50"/>
  </mergerTreeBuildMassesMethod>
  
  <mergerTreeNodeMergerMethod value="singleLevelHierarchy"/>

  
  <darkMatterProfileDMOMethod value="NFW"/>
  <darkMatterProfileConcentrationMethod value="gao2008"/>
  <darkMatterProfileMinimumConcentration value="4"/>
  <haloSpinDistributionMethod value="bett2007">
    <alpha value="2.509"/>
    <lambda0 value="0.04326"/>
  </haloSpinDistributionMethod>
  <randomSpinResetMassFactor value="2.0"/>

  
  <accretionHaloMethod value="simple">
    <redshiftReionization value="10.5"/>
    <velocitySuppressionReionization value="35.0"/>
  </accretionHaloMethod>

  
  <hotHaloMassDistributionMethod value="betaProfile"/>
  <hotHaloTemperatureProfileMethod value="virial"/>
  <hotHaloMassDistributionCoreRadiusMethod value="virialFraction">
    <coreRadiusOverVirialRadius value="0.3"/>
  </hotHaloMassDistributionCoreRadiusMethod>
  <hotHaloAngularMomentumLossFraction value="0.3"/>
  <coolingSpecificAngularMomentumMethod value="constantRotation">
    <sourceAngularMomentumSpecificMean value="hotGas"/>
    <sourceNormalizationRotation value="hotGas"/>
  </coolingSpecificAngularMomentumMethod>
  <hotHaloOutflowReincorporationMethod value="haloDynamicalTime">
    <multiplier value="5.0"/>
  </hotHaloOutflowReincorporationMethod>

  <coolingFunctionMethod value="atomicCIECloudy"/>
  <coolingRadiusMethod value="simple"/>
  <coolingRateMethod value="whiteFrenk1991">
    <velocityCutOff value="10000"/>
  </coolingRateMethod>
  <coolingTimeMethod value="simple">
    <degreesOfFreedom value="3.0"/>
  </coolingTimeMethod>
  <coolingTimeAvailableMethod value="whiteFrenk1991">
    <ageFactor value="0"/>
  </coolingTimeAvailableMethod>
  <starveSatellites value="false"/>

  
  <hotHaloRamPressureStrippingMethod value="font2008"/>
  <hotHaloRamPressureForceMethod value="font2008"/>
  <hotHaloRamPressureTimescaleMethod value="ramPressureAcceleration"/>
  <hotHaloOutflowStrippingEfficiency value="0.1"/>
  <hotHaloTrackStrippedGas value="true"/>

  
  <ramPressureStrippingDisksMethod value="null"/>
  <ramPressureStrippingSpheroidsMethod value="null"/>

  
  <tidalStrippingDisksMethod value="null"/>
  <tidalStrippingSpheroidsMethod value="null"/>
  <satelliteTidalFieldMethod value="null"/>

  
  <galacticStructureSolverMethod value="equilibrium"/>
  <darkMatterProfileMethod value="adiabaticGnedin2004">
    <A value="0.73"/>
    <omega value="0.7"/>
  </darkMatterProfileMethod>

  <spheroidAngularMomentumAtScaleRadius value="0.5"/>

  
  <galacticDynamicsBarInstabilityMethod value="efstathiou1982">
    <stabilityThresholdGaseous value="0.7"/>
    <stabilityThresholdStellar value="1.1"/>
  </galacticDynamicsBarInstabilityMethod>
  
  
  <starFormationTimescaleDisksMethod value="intgrtdSurfaceDensity"/>
  <starFormationRateSurfaceDensityDisksMethod value="krumholz2009">
    <frequencyStarFormation value="0.385"/>
    <clumpingFactorMolecularComplex value="5.000"/>
    <molecularFractionFast value="true"/>
  </starFormationRateSurfaceDensityDisksMethod>
  <starFormationTimescaleSpheroidsMethod value="dynamicalTime">
    <efficiency value="0.04"/>
    <exponentVelocity value="2.0"/>
    <timescaleMinimum value="0.001"/>
  </starFormationTimescaleSpheroidsMethod>

  
  <stellarPopulationPropertiesMethod value="instantaneous"/>
  <stellarPopulationSpectraMethod value="FSPS"/>
  <stellarPopulationSelectorMethod value="fixed"/>
  
  <initialMassFunctionMethod value="chabrier2001"/>
  <stellarPopulationMethod value="standard">
    <recycledFraction value="0.46"/>
    <metalYield value="0.035"/>
  </stellarPopulationMethod>

  
  <hotHaloExcessHeatDrivesOutflow value="true"/>
  <blackHoleHeatsHotHalo value="true"/>

  
  <starFormationFeedbackDisksMethod value="powerLaw">
    <velocityCharacteristic value="250.0"/>
    <exponent value="3.5"/>
  </starFormationFeedbackDisksMethod>
  <starFormationFeedbackSpheroidsMethod value="powerLaw">
    <velocityCharacteristic value="100.0"/>
    <exponent value="3.5"/>
  </starFormationFeedbackSpheroidsMethod>

  
  <accretionDisksMethod value="switched">
    <accretionRateThinDiskMaximum value="0.30"/>
    <accretionRateThinDiskMinimum value="0.01"/>
    <scaleADAFRadiativeEfficiency value="true"/>
    <accretionDisksShakuraSunyaevMethod value="shakuraSunyaev"/>
    <accretionDisksADAFMethod value="ADAF">
      <efficiencyRadiationType value="thinDisk"/>
      <adiabaticIndex value="1.444"/>
      <energyOption value="pureADAF"/>
      <efficiencyRadiation value="0.01"/>
      <viscosityOption value="fit"/>
    </accretionDisksADAFMethod>
  </accretionDisksMethod>

  
  <blackHoleBinaryMergersMethod value="rezzolla2008"/>
  <blackHoleSeedMass value="100"/>
  <blackHoleWindEfficiency value="0.0024"/>
  <blackHoleWindEfficiencyScalesWithRadiativeEfficiency value="true"/>
  <bondiHoyleAccretionEnhancementHotHalo value="6.0"/>
  <bondiHoyleAccretionEnhancementSpheroid value="5.0"/>
  <bondiHoyleAccretionTemperatureSpheroid value="100"/>
  <bondiHoyleAccretionHotModeOnly value="true"/>

  
  <satelliteOrbitStoreOrbitalParameters value="true"/>

  
  <virialOrbitMethod value="benson2005"/>
  <satelliteMergingTimescalesMethod value="jiang2008">
  <timescaleMultiplier value="0.75"/>
  </satelliteMergingTimescalesMethod>
  <mergerMassMovementsMethod value="simple">
  <destinationGasMinorMerger value="spheroid"/>
    <massRatioMajorMerger value="0.25"/>
  </mergerMassMovementsMethod>
  <mergerRemnantSizeMethod value="cole2000">
    <energyOrbital value="1"/>
  </mergerRemnantSizeMethod>

  
  <spheroidEnergeticOutflowMassRate value="1.0e-2"/>

  
  <mergerTreeNodeEvolverMethod value="standard">
    <odeToleranceAbsolute value="0.01"/>
    <odeToleranceRelative value="0.01"/>
  </mergerTreeNodeEvolverMethod>

  <diskMassToleranceAbsolute value="1.0e-6"/>
  <spheroidMassToleranceAbsolute value="1.0e-6"/>
  <mergerTreeEvolverMethod value="standard">
    <timestepHostAbsolute value="1.0"/>
    <timestepHostRelative value="0.1"/>
  </mergerTreeEvolverMethod>

  
  <galacticusOutputFileName value="benchmark_current.hdf5"/>
  <mergerTreeOutputterMethod value="standard">
    <outputReferences value="false"/>
  </mergerTreeOutputterMethod>

</parameters>

**Comment by abensonca** [comment]
[benchmark_current.xml](https://gist.github.com/abensonca/6a6534bc606314f21861dd54f4226fba)

**Comment by abensonca** [comment]
Some options for optimization:

- Memoization?
- Quick determination of whether the potential from the DM halo is negligible and if so ignore contraction?

**Comment by abensonca** [comment]
This is resolved by 0390db2 which allows for a change of variables when integrating to find the potential - thereby allowing the integral to be over the initial radius instead of the final radius (such that we no longer need to do root finding at each evaluating of the integrand).
